src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -818,41 +885,49 @@
           (utils/create-ca-extensions issuer-public-key ca-public-key))))
 
-(schema/defn read-infra-nodes
-    "Returns a list of infra nodes or infra node serials from the specified file organized as one item per line."
-    [infra-file-reader :- Reader]
-    (line-seq infra-file-reader))
-
-(defn- write-infra-serials-to-writer
-  [writer infra-nodes-path signeddir]
-  (try
+(schema/defn extract-active-infra-serials :- [BigInteger]
+  "Read the infra nodes file to determine which nodes are infrastructure nodes.
+  For each node, check the inventory file for any serial numbers for that node,
+  Also check the filesystem for a signed cert for that node.  Return a sorted unique
+  set of serial numbers for nodes in the infra file"
+  [{:keys [infra-nodes-path signeddir] :as settings} :- CaSettings]
+  (if (fs/exists? infra-nodes-path)
     (with-open [infra-nodes-reader (io/reader infra-nodes-path)]
       (let [infra-nodes (read-infra-nodes infra-nodes-reader)]
-        (doseq [infra-node infra-nodes]
-          (try
-            (let [infra-serial (-> (path-to-cert signeddir infra-node)
-                                   (utils/pem->cert)
-                                   (utils/get-serial))]
-              (.write writer (str infra-serial))
-              (.newLine writer))
-            (catch FileNotFoundException _
-              (log/warn
-               (i18n/trs
-                (str
-                 "Failed to find/load certificate for Puppet Infrastructure Node:"
-                 infra-node))))))))
-    (catch FileNotFoundException _
-      (log/warn (i18n/trs (str infra-nodes-path " does not exist"))))))
-
-(schema/defn generate-infra-serials
+        (loop [local-infra-nodes infra-nodes
+               result []]
+          (if-let [infra-node (first local-infra-nodes)]
+            (let [augmented-results (concat result (find-matching-valid-serial-numbers settings infra-node))
+                  cert-path (path-to-cert signeddir infra-node)]
+              (if (fs/exists? cert-path)
+                (let [serial (-> cert-path
+                                 utils/pem->cert
+                                 utils/get-serial)]
+                  (recur (rest local-infra-nodes)
+                         (conj augmented-results serial)))
+                (do
+                  (log/warn (i18n/trs "Failed to find certificate for Puppet Infrastructure Node: {0}" infra-node))
+                  (recur (rest local-infra-nodes)
+                         augmented-results))))
+            ;; ensure they are unique and ordered so the end result is stable
+            (sort (set result))))))
+    []))
+
+(schema/defn write-infra-serials-to-writer
+  [writer :- BufferedWriter
+   settings :- CaSettings]
+  (let [infra-serials (extract-active-infra-serials settings)]
+    (doseq [infra-serial infra-serials]
+      (.write writer (str infra-serial))
+      (.newLine writer))))
+
+(schema/defn generate-infra-serials!
```

Comment:
Renamed this by adding the `!` suffix to indicate that it makes changes.

Reply:
Just a sanity check here:
When a cert is renewed the new serial will be appended to the infra-serials file, but when the CA restarts, reloads, or the infra node certname file is updated the whole infra-serials file will be rewritten with the latest serial for each infra node. That's the intended behavior?

Reply:
In the startup case, with these changes to `generate-infra-serials!` we are now looking in the inventory as well for matching serial numbers and adding them.  So at startup there should be technically no work to do, unless someone manually changed the files.  The idea is that the infra-serial should always be more correct now

Reply:
So to answer your question, the intent is that we append the infra-serial on demand when a new one is generated, and when we restart, we update the state based on the current state on the filesystem, including all known valid serial numbers for the infra-nodes, not just the most recent one.  This helps at revocation time to ensure the infra-crl is updated with all possible certs for the host being revoked.

Reply:
Thanks for that.